src/sys/pthread.rs

```diff
@@ -39,4 +40,57 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
     let res = unsafe { libc::pthread_kill(thread, sig) };
     Errno::result(res).map(drop)
 }
+
+/// Value to pass with a signal. Can be either integer or pointer.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub enum SigVal {
+    /// Use this variant to pass a integer to the sigval union
+    Int(libc::c_int),
+    /// Use this variant to pass a pointer to the sigval union
+    Ptr(*mut libc::c_void),
+}
+
+// Because of macro/trait machinery in libc, libc doesn't provide this union.
```

Comment:
The problem is there already is a sigval in libc and is defined as a struct(which is used in many places since signal handling has several syscalls associated with it). That struct has PartialEq/Hash/Debug etc implemented, and has done so for a long time. There's no practical way to implement PartialEq/Hash/Debug for a union b/c that would lead to hashing/equality checking/printing uninitialized memory. So changing the current struct to a union is a breaking change. 

I can add a second definition in libc, but it would have to have a different name. If that's desired I guess reply to this comment. I'll also update that comment with the info in this one.

Reply:
Have you seen https://github.com/rust-lang/libc/pull/2813 ?  It deals with another union that was originally defined as a struct in libc.  The solution we came up with there was to define a `sigevent_0_2_126` struct identical to the old `sigevent`, and then define `sigevent` with union fields as it ought to be.  Then implement Deref and DerefMut for sigevent to sigevent_0_2_126 .  That way consumers of the old version can still access the fields that should've been in a union.  Would something similar work for sigval?